feat: prune if needed

just-mitch opened the pull request:
Fix #8572 
Fix #8608

just-mitch committed:
define escrow contract interface
implement mock escrow and include in rollup
move signature lib

just-mitch committed:
prune if needed before propose and prove.

just-mitch commented on l1-contracts/src/core/sequencer_selection/Leonidas.sol:
Ordering of functions changed to satisfy linter.

just-mitch committed:
update getSlotAt check

just-mitch committed:
Update test to hit code path

LHerskind commented on l1-contracts/src/core/interfaces/IProofCommitmentEscrow.sol:
```diff
@@ -0,0 +1,13 @@
+
+import {DataStructures} from "../libraries/DataStructures.sol";
+
+interface IProofCommitmentEscrow {
```
I don't think this interface will work, unless you also do a lot of the same accounting in here as you do in the rollup. We should do them in either one or the other. 

Specifically, look at the `stakeBond` and `unstakeBond` the stake gets a lot of information, some useful some not, but it knows who is bonding etc. But at the `unstake` it won't be told who is it that is unstaking which makes it weird here. 

In the likely case we have, where we are building the escrow specifically for the rollup, you can have the rollup provide enough information for the escrow to know what to do without doubling the accounting.

just-mitch replied:
I agree that one escrow per rollup is simpler- that is a deviation from what I was thinking.

LHerskind replied:
The `unstakeBond` is still not identifying who, so it still needs to store a bunch of storage on its own. I suggest making the "mock" without validation but still doing the transfers, such that you can more easily get an idea around the information that needs to be passed along for the minimum. 

Also, it is not enough to give a signature, to verify it we need to know the message etc. Either have that logic in the rollup, and then the escrow becomes just `bond(address _bonder, uint256 _amount) onlyRollup` or have it deal with the full thing.

LHerskind commented on l1-contracts/test/Rollup.t.sol:
```diff
@@ -113,17 +115,11 @@ contract RollupTest is DecoderBase {
     rollup.prune();
   }
 
-  function testPrune() public setUpFor("mixed_block_1") {
```
I don't like that you are using the `testBlock` to prune. We want to see that both things works. Create a separate test if you want to test a separate thing.

just-mitch committed:
feat: proofClaim in Rollup (#8636)

Fix #8608 

Add the proofClaim to the rollup.

Update the `canPrune` logic to account for it.

just-mitch commented on l1-contracts/src/core/libraries/DataStructures.sol:
```diff
@@ -79,4 +81,13 @@ library DataStructures {
+    uint256 epochToProve;
+    uint256 validUntilSlot;
+    uint256 bondAmount;
+    address rollup;
```
I'm fine to take it out, but it may create some confusion since it needs to be in the message that the prover signs over.

just-mitch committed:
clean up escrow interface
remove `TIMELINESS_PROVING_IN_SLOTS`
update tests

LHerskind commented on l1-contracts/src/core/libraries/DataStructures.sol:
```diff
@@ -79,4 +81,20 @@ library DataStructures {
     bool ignoreSignatures;
   }
+
+  struct EpochProofQuote {
```
For the `ecrecover` to be safe the message it is signing over should include the address that we expect to recover. Otherwise valid signature for a different message can also recover an address, just a different one. e.g., include the `bondProvider` in the quote.

just-mitch committed:
update escrow interface and quote structure

LHerskind reviewed (approved):
Thanks 👍 The rest can be addressed when there is logic 👍
